storage/src/main/java/tech/pegasys/teku/storage/server/DepositStorage.java

```diff
@@ -44,22 +54,69 @@ public void stop() {
   }
 
   @Override
-  public SafeFuture<Stream<DepositsFromBlockEvent>> streamDepositFromBlockEvents() {
-    return SafeFuture.completedFuture(database.streamDepositsFromBlocks());
+  public SafeFuture<ReplayDepositsResult> replayDepositEvents() {
+    return replayResult.get();
   }
 
-  @Override
-  public SafeFuture<Optional<MinGenesisTimeBlockEvent>> getMinGenesisTimeBlockEvent() {
-    return SafeFuture.completedFuture(database.getMinGenesisTimeBlock());
+  private ReplayDepositsResult replayDeposits() {
+    isSyncingFromDatabase = true;
+    final DepositSequencer depositSequencer =
+        new DepositSequencer(eth1EventsChannel, database.getMinGenesisTimeBlock());
+    try (Stream<DepositsFromBlockEvent> eventStream = database.streamDepositsFromBlocks()) {
+      eventStream.forEach(depositSequencer::depositEvent);
+    }
+    depositSequencer.depositsComplete();
+    isSyncingFromDatabase = false;
+    return depositSequencer.done();
   }
 
   @Override
   public void onDepositsFromBlock(final DepositsFromBlockEvent event) {
-    database.addDepositsFromBlockEvent(event);
+    if (!isSyncingFromDatabase) {
+      database.addDepositsFromBlockEvent(event);
+    }
   }
 
   @Override
   public void onMinGenesisTimeBlock(final MinGenesisTimeBlockEvent event) {
-    database.addMinGenesisTimeBlock(event);
+    if (!isSyncingFromDatabase) {
+      database.addMinGenesisTimeBlock(event);
+    }
+  }
+
+  static class DepositSequencer {
+    private final Eth1EventsChannel eth1EventsChannel;
+    private final Optional<MinGenesisTimeBlockEvent> genesis;
+    private boolean isGenesisDone;
+    private UnsignedLong lastDeposit;
+
+    public DepositSequencer(
+        Eth1EventsChannel eventChannel, Optional<MinGenesisTimeBlockEvent> genesis) {
+      this.eth1EventsChannel = eventChannel;
+      this.genesis = genesis;
+      this.isGenesisDone = false;
+    }
+
+    public void depositEvent(final DepositsFromBlockEvent event) {
+      if (genesis.isPresent()
+          && !isGenesisDone
+          && genesis.get().getBlockNumber().compareTo(event.getBlockNumber()) < 0) {
+        this.eth1EventsChannel.onMinGenesisTimeBlock(genesis.get());
+        isGenesisDone = true;
+      }
+      eth1EventsChannel.onDepositsFromBlock(event);
+      lastDeposit = event.getBlockNumber();
+    }
+
+    public void depositsComplete() {
+      if (genesis.isPresent() && !isGenesisDone) {
+        this.eth1EventsChannel.onMinGenesisTimeBlock(genesis.get());
+        isGenesisDone = true;
+      }
+    }
```

Comment:
This way you should not need a `depositsComplete` method. Also, it would be more in line with how we do min genesis block notification from Pow Service side. (I'm assuming `streamDepositsFromBlocks` returns `DepositsFromBlockEvent`s in order)
```suggestion
    public void depositEvent(final DepositsFromBlockEvent event) {
      eth1EventsChannel.onDepositsFromBlock(event);
      if (genesis.isPresent()
          && !isGenesisDone
          && genesis.get().getTimestamp().compareTo(event.getBlockTimestamp()) >= 0) {
        this.eth1EventsChannel.onMinGenesisTimeBlock(genesis.get());
        isGenesisDone = true;
      }
      lastDeposit = event.getBlockNumber();
    }
```

Reply:
had a chat, we'll leave it because of the possibility of minGenesis being after the last block